File: command/command/document/visitors/ListBuilder.cs
using System;
using System.IO;

namespace command.document.visitors
{
	public class ListBuilder : IDocumentContentBuilder
	{
		private readonly TextWriter _textWriter;
		private int _count = 0;

		public ListBuilder(TextWriter textWriter) => _textWriter = textWriter ?? throw new ArgumentNullException(nameof(textWriter));

		public void AfterBuild(){}
		public void BeforeBuild(){}

		public void BuildImage(IImage image)
		{
			_textWriter.WriteLine($"{_count++}. Image: {image.Width} {image.Height} {image.Resource.FilePath}");
		}

		public void BuildParagraph(IParagraph paragraph)
		{
			_textWriter.WriteLine($"{_count++}. Paragraph: {paragraph.Text}");
		}

		public void BuildTitle(string title)
		{
			_textWriter.WriteLine($"Title: {title}");
		}
	}
}

File: command_java/DocumentMenu/src/menu/action/details/ListBuilder.java
package menu.action.details;

import document.documentItem.IDocumentItemVisitor;
import document.documentItem.ItemsCollection.IDocumentItem;
import document.documentItem.image.IImage;
import document.documentItem.paragraph.IParagraph;
import document.documentItem.title.IDocumentTitle;
import menu.ILogger;

class ListBuilder implements IDocumentStateBuilder {

    public ListBuilder(ILogger logger) {
        _visitor = new ListBuilderVisitor(logger);
    }

    @Override
    public void BuildTitle(IDocumentTitle title) {
        title.Accept(_visitor);
    }

    @Override
    public void BuildDocumentItem(IDocumentItem item) {
        item.Accept(_visitor);
    }

    @Override
    public void BeforeBuild() {
    }

    @Override
    public void AfterBuild() {
    }

    private final ListBuilderVisitor _visitor;


    private class ListBuilderVisitor implements IDocumentItemVisitor {

        private ListBuilderVisitor(ILogger logger) {
            _logger = logger;
        }

        @Override
        public void Visit(IDocumentTitle title) {
            _logger.Log("Title: " + title.GetTitle());
        }

        @Override
        public void Visit(IParagraph paragraph) {
            _logger.Log(new StringBuilder().append("#").append(_count++).append(" paragraph: ").append(paragraph.GetText()).toString());
        }

        @Override
        public void Visit(IImage image) {
            _logger.Log(new StringBuilder().append("#").append(_count++).append(" image: ")
                    .append("w=").append(image.GetSize().Width()).append(", h=").append(image.GetSize().Height())
                    .append("name=").append(image.GetResource().Name()).toString());
        }

        private int _count = 0;
        private final ILogger _logger;
    }
}
